File: lib/Models/PromotionModels.dart
import 'package:android/Utils/ApiConfig.dart';
import 'BaseResponse.dart';

String? _resolvePromotionImageUrl(String? value) {
  if (value == null || value.trim().isEmpty) return null;
  return '${buildApiBasePath()}/storage/promotion/${value.trim()}';
}

// ==================== PromotionModel ====================
class PromotionModel {
  PromotionModel({
    required this.id,
    required this.code,
    required this.name,
    this.imageUrl,
    required this.discountPercent,
    required this.quantity,
    required this.remaining,
    required this.active,
    this.description,
    this.startDate,
    this.endDate,
  });

  final String id;
  final String code;
  final String name;
  final String? description;
  final String? imageUrl;
  final int discountPercent;
  final int quantity;
  final int remaining;
  final bool active;
  final DateTime? startDate;
  final DateTime? endDate;

  bool get hasStock => remaining > 0;
  String? get displayImageUrl => _resolvePromotionImageUrl(imageUrl);

  factory PromotionModel.fromJson(Map<String, dynamic> json) {
    return PromotionModel(
      id: json['id'] as String,
      code: json['code'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      imageUrl: json['imageUrl'] as String?,
      discountPercent: (json['discountPercent'] as num).toInt(),
      quantity: (json['quantity'] as num).toInt(),
      remaining: (json['remaining'] as num).toInt(),
      active: json['active'] as bool,
      startDate:
          json['startDate'] != null ? DateTime.parse(json['startDate']) : null,
      endDate:
          json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
    );
  }
}

// ==================== PromotionListResponse ====================
class PromotionListResponse extends BaseResponse {
  late List<PromotionModel> data;

  PromotionListResponse({
    required int statusCode,
    required String message,
    List<PromotionModel> data = const [],
  }) : super(statusCode: statusCode, message: message) {
    this.data = data;
  }

  factory PromotionListResponse.fromJson(Map<String, dynamic> json) {
    final raw = json['data'];
    final list = raw is List ? raw : const [];
    return PromotionListResponse(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: list
          .whereType<Map<String, dynamic>>()
          .map(PromotionModel.fromJson)
          .toList(),
    );
  }
}

File: mobile_restaurant/models/PromotionModels.ts
import { buildApiBasePath } from '../utils/ApiConfig';
import { parseBaseResponse, type BaseResponse } from './AuthModels';

function resolvePromotionImageUrl(value?: string): string | undefined {
  if (!value || !value.trim()) return undefined;
  if (/^https?:\/\//i.test(value)) return value.trim();
  return `${buildApiBasePath()}/storage/promotion/${value.trim()}`;
}

// ==================== PromotionModel ====================
export interface PromotionModel {
  id: string;
  code: string;
  name: string;
  description?: string;
  imageUrl?: string;
  discountPercent: number;
  quantity: number;
  remaining: number;
  active: boolean;
  startDate?: string;
  endDate?: string;
  hasStock: boolean;
  displayImageUrl?: string;
}

export function parsePromotionModel(json: any): PromotionModel {
  const remaining = Number(json.remaining);
  return {
    id: json.id,
    code: json.code,
    name: json.name,
    description: json.description ?? undefined,
    imageUrl: json.imageUrl ?? undefined,
    discountPercent: Number(json.discountPercent),
    quantity: Number(json.quantity),
    remaining,
    active: json.active,
    startDate: json.startDate ?? undefined,
    endDate: json.endDate ?? undefined,
    hasStock: remaining > 0,
    displayImageUrl: resolvePromotionImageUrl(json.imageUrl),
  };
}

// ==================== PromotionListResponse ====================
export interface PromotionListResponse extends BaseResponse {
  data: PromotionModel[];
}

export function parsePromotionListResponse(json: any): PromotionListResponse {
  const base = parseBaseResponse(json);
  const raw = json.data;
  const list = Array.isArray(raw) ? raw : [];
  return {
    ...base,
    data: list.map(parsePromotionModel),
  };
}
